XUtils.c

```diff
@@ -224,6 +232,209 @@ size_t String_safeStrncpy(char* restrict dest, const char* restrict src, size_t
    return i;
 }
 
+#ifdef HAVE_LIBNCURSESW
+ATTR_NONNULL_N(3)
+static size_t String_encodeWChar(void* pos, wchar_t wc, void* state) {
+   char tempBuf[MB_LEN_MAX];
+
+   size_t len = wcrtomb((pos ? (char*)pos : tempBuf), wc, (mbstate_t*)state);
+   assert(len != (size_t)-1);
+   assert(len > 0);
+
+   return len;
+}
+#else
+static size_t String_encodeWChar(void* pos, int c, void* state) {
+   (void)state;
+
+   if (pos) {
+      *(char*)pos = (char)c;
+   }
+   return 1;
+}
+#endif
+
+size_t EncodePrintableString(void* dest, const char* src, size_t srcLen, EncodeWChar encodeWChar, void* encState) {
+   assert(src || srcLen == 0);
+
+   size_t srcPos = 0;
+   size_t destSize = 0;
+   bool wasReplaced = false;
+
+#ifdef HAVE_LIBNCURSESW
+
+   const wchar_t replacementChar = CRT_utf8 ? L'\xFFFD' : L'?';
+
+   mbstate_t decState;
+   memset(&decState, 0, sizeof(decState));
+
+   mbstate_t newState;
+   wchar_t wc;
+   do {
+      wc = L'\0';
+      size_t len = 0;
+
+      bool shouldReplace = false;
+      if (srcPos < srcLen) {
+         memcpy(&newState, &decState, sizeof(newState));
+         len = mbrtowc(&wc, &src[srcPos], srcLen - srcPos, &newState);
+      }
+
+      switch (len) {
+
+      case (size_t)-2:
+         errno = EILSEQ;
+         shouldReplace = true;
+         len = srcLen - srcPos;
+         break;
+
+      case (size_t)-1:
+         shouldReplace = true;
+         len = 1;
+         break;
+
+      case 0:
+         len = 1;
+         // Fallthrough
+
+      default:
+         memcpy(&decState, &newState, sizeof(decState));
+      }
+
+      srcPos += len;
+
+      if (!shouldReplace && wc != L'\0' && !iswprint(wc)) {
+         shouldReplace = true;
+      }
+
+      if (shouldReplace) {
+         wc = replacementChar;
+         if (wasReplaced) {
+            continue;
+         }
+      }
+      wasReplaced = shouldReplace;
+
+      destSize += encodeWChar((dest ? (char*)dest + destSize : NULL), wc, encState);
```

Comment:
Treat the `destSize` as opaque, similar to how it is done in the CGroup compression code. The location needs to be part of the passed state, alongside the maximum buffer size. This avoids any pointer arithmetic on `dest` here and improves re-usability with other memory layouts.

Reply:
I'm reserved about this. `destSize` is required to be returned to the caller so that the caller can `malloc` the buffer that just fits. It would not be that "opaque" as you think.

It might be a good idea for `EncodePrintableString` to not care about the `destSize` at all and have the property and the counting all done by the callee (`encodeWChar`). The caller of  `EncodePrintableString` would have to extract the  `destSize` property from `encState` nevertheless.